### PR TITLE
Added includeStackTrace configuration option

### LINKED ISSUE #160: Added into report configuration additional flag "includeStackTrace": false
Hello. I use the reporter for a very large number of tests and I would like to be able to remove the trace stack and live only a few rows with an error description, since in correctly described tests it is not necessary to have a trace and it is clear what the error is. This will save a very large amount of space on the reporter page while scrolling.

Suggestion is to add into report configuration additional flag "includeStackTrace": false (by default it should be true) if needs to hide stack trace from the report and just live only error mismatched message.

Now i use my custom script that opens and changes html file (removes only stack trace), but it will be great to can do it through the report config.

example

Screenshot 2023-04-28 at 10 56 57

### PR DESCRIPTION
**Pull request by Hargne**
- Added includeStackTrace configuration option (fixes #160 )
- Optimized `IJestHTMLReporterConfig` type

### CODE REVIEW

File: test/htmlreporter.spec.ts [hunk 1]
```diff
@@ -182,6 +182,48 @@ describe("HTMLReporter", () => {
...
+          options: {
+            includeFailureMsg: true,
+            includeStackTrace: false,
+          },
+        });
+        const reportContent = await reporter.renderTestReportContent();
+        expect(reportContent).toBeDefined();
+        expect(reportContent!.toString().indexOf("at stack trace")).toBe(-1);
+      });
+      it.only("should keep stack trace in failure messages if set to true", async () => {
```
Comment (StefanPuia) [inline test/htmlreporter.spec.ts:198]: `.only` probably blocks the rest of the tests in the suite here

**Review by StefanPuia — COMMENTED** [review]
I've made those couple changes here
https://github.com/Hargne/jest-html-reporter/commit/e7e9f0435175d22d7cd328b03fd3f036c4b2fcd2

File: test/htmlreporter.spec.ts (hunk 1 above)
Comment (Hargne) [reply test/htmlreporter.spec.ts:198]: Perfect! I must have forgotten to remove it while testing 🙏

**Comment by StefanPuia** [comment]
took me a while to get my head around what's going on
`failureMessages` is the `error.stack` value, which means it will always have the message, stack, and `Error: ` at the beginning
`failureDetails` on the other hand will have _maybe_ always just `Error` objects: **[jest source](https://github.com/jestjs/jest/blob/bb39cb2c617a3334bf18daeca66bd87b7ccab28b/packages/jest-circus/src/utils.ts#L366)**
so we can probably have something like this here, instead of the regex replace:

```typescript
/////// htmlreporter.ts:394
// Test Failure Messages
if (
  test.failureDetails &&
  test.failureDetails.length > 0 &&
  this.getConfigValue("includeFailureMsg")
) {
  const failureMsgDiv = testResult.ele(
    "div",
    {
      class: "failureMessages",
    },
    " "
  );
  test.failureDetails.forEach((err: Error) => {
    failureMsgDiv.ele(
      "pre",
      { class: "failureMsg" },
      this.sanitizeOutput(err.message)
    );
  });
}
```

File: src/htmlreporter.ts [hunk 2]
```diff
@@ -404,13 +404,19 @@ class HTMLReporter {
...
-                  failureMsgDiv.ele(
-                    "pre",
-                    { class: "failureMsg" },
-                    this.sanitizeOutput(failureMsg)
-                  );
-                });
+                test.failureMessages
+                  .map((failureMsg) =>
+                    !this.getConfigValue("includeStackTrace")
+                      ? failureMsg.split("at")[0].trim().replace(/\n+$/, "")
```
Comment (StefanPuia) [inline src/htmlreporter.ts:410]: re: https://github.com/Hargne/jest-html-reporter/pull/161#issuecomment-1532239561

**Review by StefanPuia — CHANGES_REQUESTED** [review]
could we use failureDetails instead?

**Comment by Hargne** [comment]
@StefanPuia I get your idea. However, as far as I can see `failureDetails` is of type: `Array<unknown>`, meaning that we cannot be sure what it returns. In your example above you typed it as `Array<Error>`, but in fact it returns an array of objects:
```
[
  {
    matcherResult: {
      actual: 'something',
      expected: 'something-else,
      message: 'expect(received).toEqual(expected) // deep equality' +
        '\n' +
        'Expected: "something-else' +
        'Received: "something"',
      name: 'toEqual',
      pass: false
    }
  }
]
```
I cannot see the stack trace in there either, which is the main reason of this PR: to hide/show the stack trace. And in the end: doing this yields the same result as using `failureMessages`, so I cannot really see the benefit.

**Comment by Hargne** [comment]
Good point about the regex, and I totally agree. I will look at a more solid solution.

> For this issue specifically we'd just have to replace
> this.sanitizeOutput(err.message) with something like this.sanitizeOutput(this.getConfigValue("includeStackTrace") ? err.stack : err.message)

`failureDetails` does not have a `stack` property, only a `matcherResult` object which *may* contain a `message` without the stack trace

**Comment by StefanPuia** [comment]
That's fair enough. `failureDetails` would probably be up to the test runner implementation so it's better not to assume anything about it.

**Review by StefanPuia — APPROVED** [review]

**Pull request merged**